src/main.rs

```diff
@@ -0,0 +1,94 @@
+#![allow(non_upper_case_globals)]
+#![allow(non_camel_case_types)]
+#![allow(non_snake_case)]
+
+include!(concat!(env!("OUT_DIR"), "/cuda_driver_bindings.rs"));
+
+use anyhow;
+use std::ffi::CString;
+use tarpc::{
+    context,
+    server::{BaseChannel, Channel},
+    tokio_serde::formats::Json,
+};
+use futures_util::StreamExt;
+
+mod service;
+use crate::{service::Cuda as CudaService};
+
+// This is the type that implements the generated World trait. It is the business logic
+// and is used to start the server.
+#[derive(Clone)]
+struct RemoteCuda;
+
+#[tarpc::server]
+impl service::Cuda for RemoteCuda {
+    async fn cuInit(self, _: context::Context, flags: u32) -> u32 {
+        // This does not look right. cuInit should be called once per process.
+        // I suspect what we need to do is:
+        // 1. new connection
+        // 2. fork
+        // 3. forked server handles all requests for that connection...
+        unsafe {
+            cuInit(flags)
+        }
+    }
+    async fn cuDeviceGet(self, _: context::Context, ordinal: i32) -> (i32, u32) {
+        let mut dev: CUdevice = 0;
+        let devPtr: *mut CUdevice = &mut dev;
+
+        unsafe {
+            let res = cuDeviceGet(devPtr, ordinal);
+            (*devPtr, res)
+        }
+    }
+    async fn cuDeviceGetCount(self, _: context::Context) -> (i32, u32) {
+        let mut count: i32 = 0;
+        let countPtr: *mut i32 = &mut count;
+
+        unsafe {
+            let res = cuDeviceGetCount(countPtr);
```

Comment:
fwiw, countPtr is not safe here. From what i can tell, and this is something I am not 100% certain how to handle, the safe way to handle the stack is ensure 32767 `i32` is in use.

My google search term `cargo cffi safe usage of stack`.

References:
- https://www.reddit.com/r/rust/comments/1owhwi/on_stack_safety/
- https://doc.rust-lang.org/nomicon/ffi.html#creating-a-safe-interface
-  https://medium.com/dwelo-r-d/using-c-libraries-in-rust-13961948c72a

That said, a quick scan of those references don't show the information I remember about the 32k stack size.